### ISSUE TITLE
cucumber-chef conflicts with cucumber-nagios (net-ssh version)

### ISSUE THREAD

**Issue by c10l** [Atalanta/cucumber-chef#69]
I'm having trouble getting `cucumber-chef` to work as I can't require both `cucumber-chef` and `cucumber-nagios` together, since their dependent `net-ssh` versions conflict.

`cucumber-chef` needs `net-ssh ~> 2.2.2` while `cucumber-nagios` needs `net-ssh ~> 2.1.0` (see below).

Any way to work around that?

``` ruby
$ irb
irb(main):001:0> require 'cucumber-chef'
=> true
irb(main):002:0> require 'cucumber-nagios'
Gem::LoadError: Unable to activate cucumber-nagios-0.9.2, because net-ssh-2.2.2 conflicts with net-ssh (~> 2.1.0)
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/rubygems/specification.rb:1637:in `raise_if_conflicts'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/rubygems/specification.rb:746:in `activate'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/rubygems.rb:212:in `rescue in try_activate'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/rubygems.rb:209:in `try_activate'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/rubygems/custom_require.rb:59:in `rescue in require'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/rubygems/custom_require.rb:35:in `require'
    from (irb):2
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/bin/irb:12:in `<main>'
```

and

``` ruby
$ irb
irb(main):001:0> require 'cucumber-nagios'
=> true
irb(main):002:0> require 'cucumber-chef'
Gem::LoadError: Unable to activate chef-10.12.0, because net-ssh-2.1.4 conflicts with net-ssh (~> 2.2.2)
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/rubygems/specification.rb:1637:in `raise_if_conflicts'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/rubygems/specification.rb:746:in `activate'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/rubygems/custom_require.rb:51:in `block in require'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/rubygems/custom_require.rb:50:in `each'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/rubygems/custom_require.rb:50:in `require'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/cucumber-chef-1.0.3/lib/cucumber/chef.rb:5:in `<top (required)>'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/rubygems/custom_require.rb:55:in `require'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/rubygems/custom_require.rb:55:in `require'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/cucumber-chef-1.0.3/lib/cucumber-chef.rb:1:in `<top (required)>'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/rubygems/custom_require.rb:60:in `require'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/rubygems/custom_require.rb:60:in `rescue in require'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/rubygems/custom_require.rb:35:in `require'
    from (irb):2
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/bin/irb:12:in `<main>'
```

**Comment by zpatten** [comment]
Part of this issue is because cucumber-nagios' gem dependencies, which when last I looked, are referencing very old gems (in a few cases) and hasn't been updated in recent memory.

The other part is because we're pinning the gem dependencies at the moment.

I believe cucumber-nagios requires `2.1.x` of `net-ssh` and we want `2.2.x`.  You can try loosening the gem dependencies on either to get past this for the time being.  If you force cucumber-chef down to `net-ssh 2.1.x` I can not guarantee that cucumber-chef will function; I seem to remember a good reason for moving cucumber-chef to the latest version of `net-ssh`.

Bottom line as it stands now is cucumber-nagios is likely incompatible with cucumber-chef at the moment.

**Comment by Atalanta** [comment]
Cucumber-chef definitely won't work with net-ssh 2.1.x.

What is it that you want to do?  Is it as simple as wanting to use both tools?

**Comment by c10l** [comment]
No, I was having problems with cucumber-chef and cucumber-nagios working together, as described initially. I was really just trying to try out cucumber-chef, but stumbled on this very early.

I had given up because of that, but now I'm trying to use it again and hitting this error:

``` shell
$ cucumber-chef setup    
Provisioning cucumber-chef test lab platform.
Waiting for server........................
Instance ID: i-39a74125 ; IP Address 54.232.2.202
Platform provisioned.  Run cucumber-chef project to get started.
/usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/net-ssh-2.2.2/lib/net/ssh/transport/session.rb:66:in `initialize': Connection refused - connect(2) (Errno::ECONNREFUSED)
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/net-ssh-2.2.2/lib/net/ssh/transport/session.rb:66:in `open'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/net-ssh-2.2.2/lib/net/ssh/transport/session.rb:66:in `block in initialize'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/timeout.rb:54:in `timeout'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/1.9.1/timeout.rb:99:in `timeout'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/net-ssh-2.2.2/lib/net/ssh/transport/session.rb:66:in `initialize'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/net-ssh-2.2.2/lib/net/ssh.rb:186:in `new'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/net-ssh-2.2.2/lib/net/ssh.rb:186:in `start'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/net-ssh-multi-1.1/lib/net/ssh/multi/server.rb:185:in `new_session'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/net-ssh-multi-1.1/lib/net/ssh/multi/session.rb:483:in `next_session'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/net-ssh-multi-1.1/lib/net/ssh/multi/server.rb:138:in `session'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/net-ssh-multi-1.1/lib/net/ssh/multi/session_actions.rb:36:in `block (2 levels) in sessions'
```

Sorry for talking about an unrelated error, it's just that I'm kinda lost on the kinds of errors I'm getting.

**Comment by zpatten** [comment]
Something is blocking your connection to EC2:

`Connection refused - connect(2) (Errno::ECONNREFUSED)`

**Comment by c10l** [comment]
Hi, really sorry and thanks for the reply, but I had pasted the response from the wrong terminal window... This is the one:

``` shell
$ cucumber-chef setup 
Provisioning cucumber-chef test lab platform.
Waiting for server............................
Instance ID: i-6da54371 ; IP Address 54.232.2.169
Platform provisioned.  Run cucumber-chef project to get started.
/usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/cucumber-chef-1.0.3/lib/cucumber/chef/provisioner.rb:57:in `upload_cookbook': undefined method `upload_cookbook' for #<Chef::CookbookUploader:0x007f9934c9a0c0> (NoMethodError)
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/cucumber-chef-1.0.3/bin/cucumber-chef:61:in `setup'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/thor-0.16.0/lib/thor/task.rb:27:in `run'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/thor-0.16.0/lib/thor/invocation.rb:120:in `invoke_task'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/thor-0.16.0/lib/thor.rb:275:in `dispatch'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/thor-0.16.0/lib/thor/base.rb:425:in `start'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/lib/ruby/gems/1.9.1/gems/cucumber-chef-1.0.3/bin/cucumber-chef:123:in `<top (required)>'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/bin/cucumber-chef:23:in `load'
    from /usr/local/Cellar/rbenv/0.3.0/versions/1.9.3-p194/bin/cucumber-chef:23:in `<main>'
```

**Comment by zpatten** [comment]
Ahh, this issue was recently fixed by Stephen in commit 718960437dd1bb37c32d271019fd5cda7d080eb8.  Grab the latest version of the gem as this should be resolved in that release.

**Comment by c10l** [comment]
That did it! I'll try updating gems before banging my head against the wall next time. :)

I guess this issue could be closed, since I don't even have a way to properly test/reproduce it at the moment, but I'll leave that for you to decide.

Thanks for the help!

**Comment by Atalanta** [comment]
Depends if we feel that interoperability with Cucumber-nagios is a big deal.  I guess there are hacks around it, like gemsets, but I guess this is life the Ruby ghetto? ;)

**Comment by c10l** [comment]
Doesn't cucumber-chef use the ssh steps from cucumber-nagios? At least that's what I understood when reading the book. If it's so, then I believe that this interoperability is quite important. :)

**Comment by zpatten** [comment]
It did in 1.x; those have since been pulled directly into this project and modified to work better within Cucumber-Chef.  If you look at the cucumber-nagios repo on GitHub, it is clearly dated and movement on the project has been minimal for the past two years.  There are many other cuke gems that have essentially superseded cucumber-nagios from what I can tell and have read myself.

My advice would be to not become dependant on cucumber-nagios.

Ultimately and at first glance, IMO, the best fix for cucumber-nagios would be to losen their dependency on net-ssh (this may create a whole host of other issues thou) so it will play nice with the modern versions of the net-ssh gems.

**Comment by c10l** [comment]
I'm cool with that. Out with cucumber-nagios then, I'm closing the issue. :)
